### ISSUE TITLE
Refactoring attributes in the compiler

### ISSUE THREAD

**Issue by jdonszelmann** [rust-lang/rust#131229]
While working on #125418 with @m-ou-se, I've interacted quite a bit with attributes in the compiler. I've got some thoughts about the way they currently work. I'm posting this as a mix between an explanation of the status quo and why I think that's an issue, in addition to also serving as a kind of tracking issue for these changes if I've convinced you that this is a problem.

## Quick Overview

From the ground up: There are several syntaxes for macros, one of those syntaxes is attributes which can have [several forms]. Attributes can be expanded, either as a user defined attribute macro, or as an "active" built in attribute like `#[test]`. However, some attributes are kept around for the entire compilation lifecycle. 

These [built-in attributes] are never expanded. Instead, they are kept around and serve as markers or metadata to guide the compilation process at various stages. There are currently around `100` of these.

[several forms]: https://doc.rust-lang.org/nightly/reference/attributes.html#meta-item-attribute-syntax
[built-in attributes]: https://rustc-dev-guide.rust-lang.org/attributes.html#builtininert-attributes

## The problem

While most of what is parsed, is later lowered during [`rustc_ast_lowering`], attributes are not, mostly. Many crates under `compiler/`  depend on `rustc_ast` *just* to use `ast::Attribute`. Let's see what that means:

[`rustc_ast_lowering`]: https://github.com/rust-lang/rust/tree/master/compiler/rustc_ast_lowering

### Partial lowering and impossible states

One part of attributes actually *is* lowered, attributes of the form `#[key = "value"]` aka `MetaNameValueStr`. To be able to do that, the ast contains an enum `AttrArgsEq` that already has a variant for when eventually it is lowered:

https://github.com/rust-lang/rust/blob/11ee3a830b8537976d54805331cc626604afbb63/compiler/rustc_ast/src/ast.rs#L1697-L1700

For one part of the compilation process, the `Ast` variant is always active and `Hir` is completely unused, while later in the compiler the reverse is true. In some places people didn't realize this and they provided implementations for both cases while only one could occur,
while in other places they are marked as unreachable, like here:

https://github.com/rust-lang/rust/blob/11ee3a830b8537976d54805331cc626604afbb63/compiler/rustc_ast/src/visit.rs#L1241

Another case of partial lowering is the tokens field:

https://github.com/rust-lang/rust/blob/11ee3a830b8537976d54805331cc626604afbb63/compiler/rustc_ast_lowering/src/lib.rs#L948

Which is later extensively defended against, making sure this really happened:

https://github.com/rust-lang/rust/blob/11ee3a830b8537976d54805331cc626604afbb63/compiler/rustc_query_system/src/ich/impls_syntax.rs#L41-L54

### Parse, don't validate.

I'm a big fan of the blog post [Parse, don't validate]. Generally rust's type system makes this pattern the most obvious thing to do and it's what I teach my university students every year. However, that is exactly what we aren't doing with attributes. In [`rustc_passes/check_attr.rs`] we first validate extensively, and emit various diagnostics. However, every single attribute is later parsed again where it is needed. I started making a small overview, but `100` attributes is a lot

![Image](https://github.com/user-attachments/assets/394414b5-cd7c-484a-9dc9-e96fd3573c26)

But basically, of the first 19 attributes I looked at, 5 are `Word` attributes and trivial, a few are parsed together, but in total I've found 11 completely distinct and custom parsing logics, not reusing any parts, spread over as many files and compiler crates. 

I lied a little there, the parsing does reuse some things. For example, the attributes are turned into `MetaItem`s using common logic. However, that doesn't change the fact that attributes are effectively re-validated scattered around the compiler, and many of these places have more diagnostics of their own, that could've happened during the earlier validation. It also means that at a very late stage in the compiler, we are still dealing with parsing `TokenStream`s, something that you'd think we should abstract away a little after parsing.

An example of such custom parsing logic: 

https://github.com/rust-lang/rust/blob/11ee3a830b8537976d54805331cc626604afbb63/compiler/rustc_middle/src/ty/context.rs#L1447-L1469

[Parse, don't validate]: https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/
[`rustc_passes/check_attr.rs`]: https://github.com/rust-lang/rust/blob/11ee3a830b8537976d54805331cc626604afbb63/compiler/rustc_passes/src/check_attr.rs

### Flexibility

Finally, though I have fewer concrete examples of this, sticking to `ast::Attribute` throughout the compiler removes quite some flexibility. Everything has to fit into an `ast::Attribute`, or if it doesn't, you'd have to create more variants like `AttrArgsEq::Hir` to support something in the ast that shouldn't even be part of the ast, forcing you to add a myriad of exceptions in parts of the compiler where such an extra variant isn't relevant yet. Specifically, for #125418 we noticed this because we wanted to do some limited form of name resolution for a path stored in an attribute, which proved next to impossible. 

# Ideas

1. Lower attributes during [`rustc_ast_lowering`]. 

I've got 90% of a commit ready to do this, and it's what sparked the idea for this issue. It leads to some code duplication. I'm a little unhappy about it, because it forces a lot of changes across the entire compiler, exactly because attribute parsing now happens in so many places. However, it already means that a lot of assertions can be removed because at some part of the compiler, the fact that an `Attribute` can't have certain fields and values anymore becomes encoded in the type system. I'll open a PR for this soon, and we can discuss whether we think this is a good first step. 

What also doesn't help is that `rustc_attr` currently has logic to validate attributes, but these functions are called in wildly different parts of the compiler. Some functions here validate actual `ast::Attribute`s from before lowering, while other functions validate new `hir::Attribute`s. Bugs here seem easy to make, since even though currently these are the same type, they don't always contain the same fields....

2.  The "real solution": parse, don't validate

As I see it, what would make attributes so much nicer to work with, is if there was a place in the compiler (something like the `rustc_attr` crate, but actually good) where all attributes are turned from their ast tokeny representation into some specific attribute representation. Something like the following, based on the examples I've looked at in the table I showed a little higher up:

```rust
enum InlineKind {
    Always,
    Never,
    Normal
}

enum Attribute {
    Diagnostic {
        message: Symbol,
        name: Symbol,
        notes: Vec<Symbol>
    },
    Inline(InlineKind),
    Coverage(bool),
    // ...
}
```

This structure contains only the information necessary to use each attribute, and all the diagnostics happen while parsing into this structure. That has the added benefit that this datastructure itself serves as great documentation as to what values an attribute allows. It's super clear here that a `#[diagnostic]` attributes contains a message. Currently, you'd have to make sure the written documentation for this attribute is up-to-date enough.

The translation from `ast::Attribute` to this new parsed attribute should, I think, happen during AST to HIR lowering.

I think the advantages of this should be pretty obvious, based on the examples I've given of the problems with the current approach. However, I could think of some potential blockers people might care about:

* some errors might now be thrown in different stages of the compiler (earlier). I'd say this can actually be an advantage, but I've not talked to enough people to know whether this is a problem anywhere
* A file with code to parse these attributes will contain code of many different features. I personally like that, but it also means that those features themselves become slightly less self-contained.
* Validity of attributes still needs to be checked on-site. (like `track_caller` not being valid on closures, given certain feature flags)
* Affects large parts of the compiler, also unstable parts, where people are actively opening merge requests and might run into conflicts.

Part two I have not worked on personally. I might, if I find enough time, but if someone feels very inspired to pick this up or lead this (or tell my why this is a dumb idea) feel free to.

**Comment by jieyouxu** [comment]
Zulip thread: https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/Attribute.20handling.20reworks

**Comment by bjorn3** [comment]
How will this work with custom tool attributes? These are registered using an attribute inside the crate that uses the attribute on it'w own code.

**Comment by jdonszelmann** [comment]
hmm, that's a good question. I guess those specifically we couldn't parse and they should stay around as either `ast::Attribute` or some kind of lowered `hir::Attribute` a bit like how I describe in the first part of the ideas section.

**Comment by jdonszelmann** [comment]
Related: https://github.com/rust-lang/rust/issues/131801

**Comment by safinaskar** [comment]
I totally agree with "Parse, not validate" part. There are a lot of places in compiler, where we know statically that particular enum variant is forbidden. Ideally, we should use pattern types for enums ( https://github.com/rust-lang/rust/issues/123646 ) for all such places

**Comment by jdonszelmann** [comment]
Alright, let's have an update. I'll stop updating the initial comment since most of that work is actually done now! let's build a list of what attributes have and don't have new parsers so we can keep track of the effort.

**Comment by jdonszelmann** [comment]
O

- `#[diagnostic::do_not_recommend]` @mejrs expressed interest
- `#[diagnostic::on_unimplemented]` @mejrs expressed interest, might be nice to merge with the other diagnostics attributes
- `#[coverage]`
- `#[no_sanitize]`
- `#[non_exhaustive]`
- `#[marker]`
- `#[target_feature]`
- `#[thread_local]`
- `#[track_caller]`
- `#[doc]` doc comments are done, but not the rest of the doc attribute uses  like doc hidden
- `#[no_link]`
- `#[export_name]`
- `#[rustc_layout_scalar_valid_range_start]`
- `#[rustc_layout_scalar_valid_range_end]`
- `#[debugger_visualizer]`
- `#[rustc_std_internal_symbol]`
- `#[rustc_as_ptr]`
- `#[rustc_no_implicit_autorefs]`
- `#[rustc_never_returns_null_ptr]`
- `#[rustc_legacy_const_generics]`
- `#[rustc_lint_query_instability]`
- `#[rustc_lint_untracked_query_information]`
- `#[rustc_lint_diagnostics]`
- `#[rustc_lint_opt_ty]`
- `#[rustc_lint_opt_deny_field_access]`
- `#[rustc_clean]`
- `#[rustc_dirty]`
- `#[rustc_if_this_changed]`
- `#[rustc_then_this_would_need]`
- `#[rustc_coinductive]`
- `#[rustc_must_implement_one_of]`
- `#[rustc_deny_explicit_impl]`
- `#[rustc_do_not_implement_via_object]`
- `#[const_trait]`
- `#[collapse_debuginfo]`
- `#[must_not_suspend]`
- `#[must_use]`
- `#[may_dangle]`
- `#[rustc_pass_by_value]`
- `#[rustc_allow_incoherent_impl]`
- `#[rustc_has_incoherent_inherent_impls]`
- `#[ffi_pure]` merge with ffi_const?
- `#[ffi_const]` merge with ffi_pure?
- `#[link_ordinal]`
- `#[link]`
- `#[link_name]`
- `#[link_section]`
- `#[no_mangle]`
- `#[macro_use]`
- `#[macro_escape]`
- `#[path]`
- `#[macro_export]`
- `#[ignore]`
- `#[should_panic]`
- `#[automatically_derived]`
- `#[no_implicit_prelude]`
- `#[rustc_object_lifetime_default]`
- `#[proc_macro]`
- `#[proc_macro_attribute]`
- `#[proc_macro_derive]`
- `#[autodiff_forward]`
- `#[autodiff_reverse]`
- `#[coroutine]`
- `#[type_const]`
- `#[linkage]`
- `#[rustc_pub_transparent]`
- `#[allow]` merge with expect, warn, deny, forbid
- `#[expect]` merge with allow, warn, deny, forbid
- `#[warn]` merge with allow, expect, deny, forbid
- `#[deny]` merge with allow, expect, warn, forbid
- `#[forbid]` merge with allow, expect, warn, deny
- `#[cfg]` likely complicated, I have ideas about this
- `#[cfg_attr]` idem
- `#[cfg_trace]`
- `#[cfg_attr_trace]`
- `#[export_stable]`
- `#[cfi_encoding]`
- `#[pointee]`
- `#[omit_gdb_pretty_printer_section]`
- `#[used]`
- `#[instruction_set]`
- `#[windows_subsystem]`
- `#[patchable_function_entry]`
- `#[deprecated_safe]`
- `#[prelude_import]`
- `#[panic_handler]`
- `#[allow_internal_unsafe]`
- `#[fundamental]`
- `#[lang]`
- `#[needs_allocator]`
- `#[default_lib_allocator]`
- `#[custom_mir]`

I am aware that this list is not quite complete. It does not include `#[recursion_limit]` for example.
There are likely more that aren't checked in [check_attr.rs](https://github.com/rust-lang/rust/blob/0d6ab209c525e276cbe7544cbd39a3c3619b6b18/compiler/rustc_passes/src/check_attr.rs#L120) which at this point I have no  exhaustive list of.

I've added some comments on how I think some attributes should be implemented, but those are not exhaustive either at this point.

Finally, convertin an attribute means it gets a variant here:

https://github.com/rust-lang/rust/blob/0d6ab209c525e276cbe7544cbd39a3c3619b6b18/compiler/rustc_attr_data_structures/src/attributes.rs#L153

And a parser in `rustc_attr_parsing`. That crate is pretty well documented. However, there's no proper guide yet on how to convert an attribute so for now you can contact me on zulip if any part isn't obvious. If you're done, I don't mind reviewing so feel free to assign me.

# In Progress

- `#[cold]` by jana
- `#[naked]` by jana
- [`#[crate_name]`](https://github.com/rust-lang/rust/pull/137729) by jana
- [`#[optimize]`](https://github.com/rust-lang/rust/pull/138291) by jana
- [`#[inline]`](https://github.com/rust-lang/rust/pull/138165) by jana

# Done:

- `#[rustc_allow_const_fn_unstable]`
- `#[allow_internal_unstable]`
- `#[rustc_default_body_unstable]`
- `#[rustc_confusables]`
- `#[rustc_const_stable] & #[rustc_const_unstable]`
- `#[rustc_const_stable_indirect]`
- `#[deprecated]`
- `#[doc = ..]` (only doc comments, not other doc attributes)
- `#[rustc_macro_transparency]`.
- `#[repr]`
- `#[stable]`, `#[unstable]` & `#[rustc_allowed_through_unstable_modules]`

See also: https://github.com/rust-lang/rust/blob/master/compiler/rustc_attr_data_structures/src/attributes.rs#L153

If you think an update to this list is in order, feel free to contact me.

**Comment by jdonszelmann** [comment]
cc: @ehuss

**Comment by jdonszelmann** [comment]
The time has come that most attributes have been converted to new parsers (see https://github.com/rust-lang/rust/issues/131229#issuecomment-2971351163). So, what will we do now? Some open issues are:

- [ ] Reduce the size of `check_attrs.rs`, move as much as possible into the attribute parsers. **(ready to do)** (claimed by @JayanAXHF)
- [ ] Filter out derive helper attributes when lowering, otherwise they'd show up as unparsed attributes too **(ready to do)**
- [x] Simplify `find_attr!()` for the case when a hir id is given https://github.com/rust-lang/rust/pull/153582
- [ ] Automatically generate templates? **(ready to do)**
- [x] Clean up diagnostic attributes (@mejrs)
- [ ] Create iterators over all the spans in attributes (claimed by jana) 
- [ ] Allow representing subsets of the large `AttributeKind` enum like for the cross-crate encodable ones (claimed by jana)
- [ ] Get rid of the "unparsed" variant in hir attributes. Instead we should call it "tool" as only tool attributes can be unparsed **(blocked on the last attr ports)**
- [ ] Remove `AttributeExt` **(blocked on the last attr ports)**
- [ ] Remove `rustc_ast::attr` **(blocked on the last attr ports)**
- [ ] Remove `builtin_attrs.rs` **(blocked on the last attr ports)**
- [ ] Ensure we can never ICE because of early parsings (like on macros)  **(blocked on the last attr ports)**
- [x] rename `unexpected_literal` to be positive
- [ ] remove visit_attr and only keep visit_attrs
- [ ] remove TLS for parsers

**Comment by kantnero** [comment]
Hi @jdonszelmann,
while #157896 is still being reviewed I'd like to take up another issue from the task listed here if you don't mind. Something easy-medium perhaps?

**Comment by mejrs** [comment]
I wouldn't say any of these are easy. Maybe you could help with documenting attributes in https://github.com/rust-lang/rust/issues/157604?

**Comment by kantnero** [comment]
Sure, I don't mind
